File: internal/advisor/advisor.go
// Package advisor answers natural-language questions about a server's pals
// and base crews through a hosted model API — Anthropic's Claude
// (claude.go) or Google's Gemini (gemini.go), whichever key the operator
// configured. Both speak the same Chat shape, so the HTTP handler and the
// frontend never know which one is behind it.
//
// The split with the frontend is deliberate: every derived number
// (effective work levels, condenser math, stat estimates) lives in the
// browser's calculators, and the vendored game catalogs never made it into
// the Go binary. So the browser builds a compact JSON summary of what it
// already computed and sends it with each question; this package only wraps
// it in a system prompt that explains the schema and the game's mechanics,
// and keeps the API key on the server where the browser can never see it.
package advisor

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"time"
)

// Providers, as spelled in config, the store, and the status endpoint.
const (
	ProviderAnthropic = "anthropic"
	ProviderGemini    = "gemini"
)

// ModelOption is one model a key's owner may pick. A curated list rather
// than free text: a mistyped model ID would sit in the store and 404 on
// every question. The first entry per provider is the default.
type ModelOption struct {
	ID    string `json:"id"`
	Label string `json:"label"`
}

// ModelOptions rides the status payload so the UI's picker and this
// package's validation can never drift. Kept to generally-available models
// on purpose — preview IDs get retired on a schedule and take the advisor
// down with them (gemini-3-pro-preview did exactly that, 2026-03).
var ModelOptions = map[string][]ModelOption{
	ProviderAnthropic: {
		{ID: "claude-opus-5", Label: "Claude Opus 5 · most capable"},
		{ID: "claude-sonnet-5", Label: "Claude Sonnet 5 · balanced"},
		{ID: "claude-haiku-4-5", Label: "Claude Haiku 4.5 · fastest"},
	},
	ProviderGemini: {
		{ID: "gemini-3.5-flash", Label: "Gemini 3.5 Flash · most capable"},
		{ID: "gemini-3.6-flash", Label: "Gemini 3.6 Flash · efficient"},
		{ID: "gemini-3.5-flash-lite", Label: "Gemini 3.5 Flash-Lite · fastest"},
	},
}

// DefaultModel is what an empty model choice (older stored keys, env keys)
// resolves to.
func DefaultModel(provider string) string {
	if opts := ModelOptions[provider]; len(opts) > 0 {
		return opts[0].ID
	}
	return ""
}

// ValidModel reports whether a stored choice is one this package will run.
// Empty is valid — it means the provider default.
func ValidModel(provider, model string) bool {
	if model == "" {
		return true
	}
	for _, opt := range ModelOptions[provider] {
		if opt.ID == model {
			return true
		}
	}
	return false
}

// New builds the client for a provider name — the one place the string
// becomes a concrete implementation, shared by env wiring and the UI-key
// endpoint so the two can never drift on which names are valid. model may
// be empty (the provider default) but must otherwise be a ModelOptions ID.
func New(ctx context.Context, provider, apiKey, model string) (Client, error) {
	if !ValidModel(provider, model) {
		return nil, fmt.Errorf("unknown model %q for provider %q", model, provider)
	}
	switch provider {
	case ProviderAnthropic:
		return NewClaude(apiKey, model), nil
	case ProviderGemini:
		return NewGemini(ctx, apiKey, model)
	default:
		return nil, fmt.Errorf("unknown advisor provider %q", provider)
	}
}

// Client is what both providers implement; the API layer's AdvisorClient
// interface matches it structurally.
type Client interface {
	// Chat sends one model turn. asker is the signed-in palcon username —
	// injected server-side so the model's "who am I advising" guess works
	// from a name the browser can't spoof.
	Chat(ctx context.Context, asker, gameContext string, tools []Tool, history []Message) (Reply, error)
	Provider() string
	// Model is the resolved model ID questions run on, for the status UI.
	Model() string
}

// askerBlock renders the identity line both providers append to the system
// context. Empty asker (shouldn't happen behind auth) renders nothing.
func askerBlock(asker string) string {
	if asker == "" {
		return ""
	}
	return fmt.Sprintf("The person asking is signed in to palcon with the username %q.", asker)
}

// Message is one prior turn of the conversation, resent in full on every
// request — the server stores nothing between questions.
//
// The tool fields exist because the calculators the advisor can call live
// in the browser, not here: the model asks for a call (an assistant turn
// carrying toolCalls), the browser runs the real calculator and answers (a
// "tool" turn carrying toolResults), and the loop driver is the browser.
// This package only translates the exchange into each provider's wire
// format.
type Message struct {
	Role        string       `json:"role"`
	Content     string       `json:"content,omitempty"`
	ToolCalls   []ToolCall   `json:"toolCalls,omitempty"`
	ToolResults []ToolResult `json:"toolResults,omitempty"`
}

// Tool is a browser-implemented function the model may call. Definitions
// ride in from the browser with each request — the single source of truth
// stays next to the code that executes them (web/src/lib/advisor-tools.ts),
// and the server never needs to know what a breeding table is.
type Tool struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	InputSchema json.RawMessage `json:"inputSchema"`
}

// ToolCall is the model asking the browser to run one tool.
type ToolCall struct {
	ID   string          `json:"id,omitempty"`
	Name string          `json:"name"`
	Args json.RawMessage `json:"args"`
	// Signature is Gemini's thought signature for the call — opaque bytes
	// (base64 on the wire) the provider requires echoed back verbatim with
	// the call, or the next request 400s. The browser never reads it, it
	// just rides the round trip; Claude has no equivalent and ignores it.
	Signature []byte `json:"signature,omitempty"`
}

// ToolResult is the browser's answer to one call.
type ToolResult struct {
	ID      string `json:"id,omitempty"`
	Name    string `json:"name"`
	Content string `json:"content"`
}

// Reply is one model turn: final text, or a request for tool calls (Text
// may carry preamble either way).
type Reply struct {
	Text      string     `json:"text"`
	ToolCalls []ToolCall `json:"toolCalls,omitempty"`
}

// ErrRefused reports that the model declined to answer. Its own message when
// surfaced to the user, rather than a generic upstream error, because a
// retry won't help and the question is what has to change.
var ErrRefused = errors.New("the advisor declined to answer this question")

// RateLimitedError reports the provider refusing for quota or rate reasons.
// Its own type rather than a generic upstream error because it's actionable
// for whoever owns the key — wait it out, or fix the plan — and the handler
// words the message differently for a personal key than a shared one.
type RateLimitedError struct {
	// RetryAfter is the provider's suggested wait, 0 when it named none.
	RetryAfter time.Duration
}

func (e *RateLimitedError) Error() string {
	if e.RetryAfter > 0 {
		return fmt.Sprintf("rate limited, retry in %s", e.RetryAfter.Round(time.Second))
	}
	return "rate limited"
}

// Both providers spell the suggested wait inside the message text ("Please
// retry in 57.05s") more reliably than in typed fields; scrape it, treat a
// miss as unknown.
var retryAfterRe = regexp.MustCompile(`(?i)retry in ([0-9]+(?:\.[0-9]+)?)\s*s`)

func rateLimited(message string) *RateLimitedError {
	if m := retryAfterRe.FindStringSubmatch(message); m != nil {
		if secs, err := strconv.ParseFloat(m[1], 64); err == nil {
			return &RateLimitedError{RetryAfter: time.Duration(secs * float64(time.Second))}
		}
	}
	return &RateLimitedError{}
}

// systemPrompt teaches the model its role, the shape of the data the browser
// sends, and the handful of game mechanics its recommendations turn on. The
// mechanics are restated here even though the numbers in the data already
// include them, so the model explains *why* (condense, souls, books) instead
// of just reading levels back. Shared verbatim by both providers so a server
// switching between them gives the same kind of advice.
const systemPrompt = `You are the advisor inside palcon, a management console for Palworld dedicated servers. You help players and admins with anything Palworld or palcon: what to do with their pals and base crews, general game questions (species, items, mechanics, bosses, locations), and how to use the console itself. You appear as a small chat on every page.

Your grounding, in order of preference:
1. The server's actual save data, which arrives in a <server_data> JSON block — use it for anything about this server's pals, players and bases.
2. Your tools: the console's own calculators (breeding, inheritance, stats), search_palcon_docs for how palcon itself works, and palworld_wiki for game facts you aren't certain of.
3. General knowledge, for well-known game basics only.

The <server_data> was computed by the console's own calculators (it can be sparse or empty when the server has no save configured — you can still answer game and console questions):
- "bases" each carry a work board: for the 12 work suitabilities, the best operational level on the crew, who covers it, and (when someone in the guild's boxes would do the job better) an "upgrade" naming that pal. Levels are effective: species base + work books + condenser star bonus, capped at 10, plus a non-stacking +1 when a deployed partner-skill buffer covers that type.
- Each base lists its crew (with per-pal top suitabilities, sickness, switched-off jobs), total appetite, night workers, sick pals, idle pals (suited to no work), and partner-skill buffs.
- "players" summarize each player's notable pals: combat power estimates, talents (IVs, 0-100 each), condenser stars (0-4), soul ranks, passives, and where the pal sits (party, palbox, base).
- "duplicates" counts spare same-species pals available as condenser fodder.

Game mechanics your advice turns on:
- Condensing consumes duplicates of the same species: each star adds +5% to all stats, and stars 1-3 each add +1 work rank to the pal's best suitabilities (cycling best-first), with 4 stars adding +1 to every suitability it has. Recommend condensing when duplicates exist, into the copy with the best talents/passives.
- Pal Souls add +3% per rank per stat (HP, Attack, Defense are upgraded separately at the statue).
- Work books add permanent ranks to one suitability; sick pals stop working entirely; a switched-off job means the pal has the level but won't take the work.
- Talents (IVs) and passives pass to bred offspring, so a poorly-rolled workhorse is often better replaced through breeding than invested in.

Tools: use them for anything they can answer instead of guessing or declining. The calculator tools (breed_child, parents_for, inheritance_odds, estimate_stats) come from the same tables the game uses — chain calls when needed (find parent pairs, then check which of them the players own). Use search_palcon_docs when asked how palcon works (features, visibility, backups, agents, setup). Use palworld_wiki for game facts beyond the calculators — drops, locations, items, bosses — especially anything that may have changed in a game update.

Who you're advising: each request names the signed-in palcon user. Match that username against the player names in <server_data> — case-insensitively, allowing obvious variants (a tag or numbers around the same name). If exactly one player plausibly matches, assume they are that player, tailor personal recommendations ("my pals", "my base") to that player's pals, and mention the assumption once ("Going by your username I'll assume you're Aster — say so if not"). If no player matches or several could, ask which player they are before giving personal advice, and remember their answer for the rest of the conversation. Server-wide and general game questions need no identification — just answer.

How to answer:
- Lead with the recommendation or answer, then the reason. Name pals specifically: nickname (or species) and owner.
- Ground claims in the data or tool results. If neither shows something, say so plainly instead of guessing.
- Plain text only: no markdown headings, tables, bold, or emoji. Short paragraphs; use simple "- " lists when listing.
- Keep responses focused and brief. Answer what was asked; offer at most one follow-up suggestion.`

File: palcon/web/src/lib/advisor.ts
import type { Guild } from "./api";
import { palKey, palName, passiveName } from "./paldex";
import { crewReport, dedupeSpecies, topWork, WORK_TYPES, type CrewPal } from "./crew";
import { palEffectiveStats, powerScore } from "./stats";
import { nearestLandmark } from "./pois";

/**
 * Context builder for the pal advisor: the compact JSON summary the browser
 * sends with each question. Built here, not on the server, because every
 * derived number (effective work levels, condenser math, stat estimates)
 * already lives in these client-side calculators — the Go side never learned
 * the vendored catalogs, and the advisor shouldn't be the reason it has to.
 *
 * Built from the same /pals payload the page renders, so player-visibility
 * hides are already applied: what a hidden player keeps off the screen also
 * stays out of the prompt.
 */

/** The slice of the calculators' SavePal the builder reads. */
export type AdvisorPal = CrewPal & {
  nickname: string;
  level: number;
  ivHp: number;
  ivAttack: number;
  ivDefense: number;
  condenser: number;
  souls: { hp: number; attack: number; defense: number };
  passives: string[];
  isAlpha: boolean;
  trust: number;
};

export interface AdvisorContext {
  /** The JSON block the server wraps in <server_data> for the model. */
  json: string;
  /** For the "what the advisor sees" caption. */
  counts: { pals: number; bases: number; players: number };
  /** Questions derived from what the save actually shows — sick crews,
   * available swaps, condenser fodder — so the empty state invites with the
   * player's own problems instead of canned examples. */
  suggestions: string[];
}

/** A pal named the way the advisor should say it back: nickname first,
 * species in parentheses when a nickname hides it. */
function displayName(p: AdvisorPal): string {
  const species = palName(p.characterId);
  return p.nickname && p.nickname !== species ? `${p.nickname} (${species})` : species;
}

function jobLabel(type: string): string {
  return WORK_TYPES.find((w) => w.id === type)?.label ?? type;
}

export function buildAdvisorContext(pals: AdvisorPal[], guilds: Guild[]): AdvisorContext {
  // Signals the suggestion builder turns into questions afterwards — the
  // loud problems (sickness, open upgrades, uncovered nights, idle hands)
  // noted here as the boards are computed.
  const sickBases: string[] = [];
  const openUpgrades: { name: string; job: string; base: string }[] = [];
  const nightlessBases: string[] = [];
  const idleNames: string[] = [];

  // Bases, labeled the way the Crew tab labels them so the advisor and the
  // board the player is looking at use the same names.
  const baseDefs: { id: string; label: string; guildName: string; memberUids: Set<string> | null }[] = [];
  const known = new Set<string>();
  for (const g of guilds) {
    const uids = new Set(g.members.map((m) => m.uid));
    g.bases.forEach((b, i) => {
      known.add(b.id);
      const mark = nearestLandmark(b.x, b.y);
      baseDefs.push({
        id: b.id,
        label: `${g.name || "Unnamed guild"} · Base ${i + 1}` + (mark ? ` · near ${mark.name}` : ""),
        guildName: g.name || "Unnamed guild",
        memberUids: uids,
      });
    });
  }
  for (const id of new Set(pals.map((p) => p.pal.baseId).filter((b) => b && !known.has(b)))) {
    baseDefs.push({ id, label: "Unlisted base", guildName: "", memberUids: null });
  }

  const bases = baseDefs.map((def) => {
    const crew = pals.filter((p) => p.pal.baseId === def.id);
    const boxes = pals.filter(
      (p) => p.pal.baseId === "" && (def.memberUids === null || def.memberUids.has(p.playerUid)),
    );
    const report = crewReport(crew, boxes);

    const baseShort = def.label.split(" · near ")[0];
    if (report.sick.length > 0) sickBases.push(baseShort);
    const bestUpgrade = report.rows
      .filter((r) => r.upgrade)
      .sort((a, b) => b.upgrade!.level - b.upgrade!.over - (a.upgrade!.level - a.upgrade!.over))[0];
    if (bestUpgrade) {
      openUpgrades.push({
        name: displayName(bestUpgrade.upgrade!.pal as AdvisorPal),
        job: bestUpgrade.label,
        base: baseShort,
      });
    }
    if (crew.length > 0 && report.nightHands === 0) nightlessBases.push(baseShort);
    for (const p of report.idle) idleNames.push(displayName(p as AdvisorPal));

    return {
      base: def.label,
      guild: def.guildName || undefined,
      crewSize: crew.length,
      appetite: report.appetite,
      nightWorkers: report.nightHands,
      sick: report.sick.map((p) => `${displayName(p as AdvisorPal)} (${(p as AdvisorPal).pal.sick})`),
      idle: report.idle.map((p) => displayName(p as AdvisorPal)),
      ranchers: report.ranchers.map((p) => displayName(p as AdvisorPal)),
      buffs: report.buffs.map((b) => `${displayName(b.pal as AdvisorPal)} gives +1 ${jobLabel(b.type)} base-wide`),
      board: report.rows
        .filter((r) => r.best > 0 || r.upgrade)
        .map((r) => ({
          job: r.label,
          best: r.best,
          night: r.night || undefined,
          hands: dedupeSpecies(r.hands)
            .slice(0, 6)
            .map(({ pal, count }) => (count > 1 ? `${palName(pal.characterId)} x${count}` : palName(pal.characterId))),
          upgrade: r.upgrade
            ? {
                swapIn: displayName(r.upgrade.pal as AdvisorPal),
                owner: r.upgrade.pal.playerName,
                from: r.upgrade.pal.where,
                level: r.upgrade.level,
                currentBest: r.upgrade.over,
              }
            : undefined,
        })),
      crew: crew.map((p) => ({
        name: displayName(p),
        owner: p.playerName,
        level: p.level,
        stars: p.condenser || undefined,
        sick: p.pal.sick || undefined,
        work: topWork(p.pal).map((w) => `${w.label} ${w.level}${w.off ? " (switched off)" : ""}`),
      })),
    };
  });

  // Per-player rollup with their strongest pals — soul/condense/breeding
  // questions are about individuals, so the strongest carry full detail.
  const byPlayer = new Map<string, AdvisorPal[]>();
  for (const p of pals) {
    const list = byPlayer.get(p.playerUid) ?? [];
    list.push(p);
    byPlayer.set(p.playerUid, list);
  }
  const players = [...byPlayer.values()].map((list) => {
    const notable = list
      .map((p) => ({ p, stats: palEffectiveStats(p.pal) }))
      .sort((a, b) => (b.stats ? powerScore(b.stats) : 0) - (a.stats ? powerScore(a.stats) : 0))
      .slice(0, 10)
      .map(({ p, stats }) => ({
        name: displayName(p),
        where: p.where,
        level: p.level,
        talents: `${p.ivHp}/${p.ivAttack}/${p.ivDefense}`,
        stars: p.condenser,
        souls: p.souls.hp + p.souls.attack + p.souls.defense > 0 ? p.souls : undefined,
        trust: p.trust,
        passives: p.passives.map(passiveName),
        alpha: p.isAlpha || undefined,
        power: stats ? Math.round(powerScore(stats)) : undefined,
      }));
    return {
      player: list[0].playerName,
      pals: list.length,
      atBases: list.filter((p) => p.pal.baseId !== "").length,
      strongest: notable,
    };
  });

  // Spare same-species copies are condenser fodder; party and base pals are
  // in use, so only boxed copies count as spare.
  const species = new Map<string, { total: number; spares: number }>();
  for (const p of pals) {
    const k = palKey(p.characterId);
    const entry = species.get(k) ?? { total: 0, spares: 0 };
    entry.total += 1;
    if (p.pal.baseId === "" && p.where !== "Party") entry.spares += 1;
    species.set(k, entry);
  }
  const duplicates = [...species.entries()]
    .filter(([, v]) => v.total > 1 && v.spares > 0)
    .sort((a, b) => b[1].spares - a[1].spares)
    .slice(0, 40)
    .map(([k, v]) => ({ species: palName(k), total: v.total, spareInBoxes: v.spares }));

  // Suggestions: one guaranteed chip for the loudest problem (a sick crew),
  // then a random draw across distinct categories so the invitations read
  // like different kinds of questions, not four spellings of one. Random on
  // purpose — the pool re-rolls per page load, which keeps the chips fresh
  // for someone who opens the chat every day.
  const pick = <T,>(arr: T[]): T => arr[Math.floor(Math.random() * arr.length)];
  const suggestions: string[] = [];
  if (sickBases.length > 0) suggestions.push(`Why have pals stopped working at ${pick(sickBases)}?`);

  const pools: string[][] = [];
  if (openUpgrades.length > 0) {
    const up = pick(openUpgrades);
    pools.push([
      `Should I swap ${up.name} into ${up.job} at ${up.base}?`,
      `Who should cover ${up.job} at ${up.base}?`,
    ]);
  }
  if (nightlessBases.length > 0) {
    pools.push([`Nobody works nights at ${pick(nightlessBases)} — who should I add?`]);
  }
  if (idleNames.length > 0) {
    pools.push([`${pick(idleNames)} isn't suited to any work — what should I do with them?`]);
  }
  if (duplicates.length > 0) {
    const dup = duplicates[0];
    pools.push(
      dup.spareInBoxes >= 4
        ? [`What should I condense first?`, `Are my ${dup.spareInBoxes} spare ${dup.species} worth condensing?`]
        : [`Is ${dup.species} worth collecting more of to condense?`],
    );
    pools.push([`What does ${dup.species} drop?`, `Where do wild ${dup.species} spawn?`]);
  }
  pools.push(["Which pals deserve my Pal Souls?", "Where should my soul upgrades go?"]);
  pools.push(["What's worth breeding next?", "How do I breed stronger workers?"]);
  pools.push(["What should I focus on next?", "How do I level my pals faster?"]);
  pools.push(["What can palcon's calculators do?", "How does the breeding path finder work?"]);

  // Shuffle the categories, then take one question from each until four
  // chips stand — every draw a different kind of ask.
  pools.sort(() => Math.random() - 0.5);
  for (const pool of pools) {
    if (suggestions.length >= 4) break;
    suggestions.push(pick(pool));
  }

  return {
    json: JSON.stringify({ bases, players, duplicates }),
    counts: { pals: pals.length, bases: bases.length, players: players.length },
    suggestions,
  };
}
